src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3600,34 +3634,200 @@ int RGWRados::rewrite_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, c
                        attrset, 0, real_time(), NULL, dpp, y);
 }
 
-int RGWRados::reindex_obj(const RGWBucketInfo& bucket_info,
+
+int RGWRados::reindex_obj(rgw::sal::Driver* driver,
+			  RGWBucketInfo& bucket_info, // NB: could this be made const?
 			  const rgw_obj& obj,
 			  const DoutPrefixProvider* dpp,
 			  optional_yield y)
 {
-  if (bucket_info.versioned()) {
-    ldpp_dout(dpp, 10) << "WARNING: " << __func__ <<
-      ": cannot process versioned bucket \"" <<
-      bucket_info.bucket.get_key() << "\"" <<
-      dendl;
-    return -ENOTSUP;
+  int ret;
+  RGWObjectCtx obj_ctx(driver);
+  RGWRados::Object rados_obj(this, bucket_info, obj_ctx, obj);
+
+  RGWRados::Object::Stat stat_op(&rados_obj);
+  ret = stat_op.stat_async(dpp);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": unable to initiate async stat of \"" << obj << "\"; returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
   }
 
-  Bucket target(this, bucket_info);
-  RGWRados::Bucket::UpdateIndex update_idx(&target, obj);
-  const std::string* no_write_tag = nullptr;
+  librados::IoCtx head_obj_ctx;
+  ret = get_obj_head_ioctx(dpp, bucket_info, obj, &head_obj_ctx);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": get_obj_head_ioctx for \"" << obj << "\" returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
+  }
 
-  int ret = update_idx.prepare(dpp, RGWModifyOp::CLS_RGW_OP_ADD, no_write_tag, y);
+  const uint64_t epoch = head_obj_ctx.get_last_version();
```

Comment:
I checked and you're right. I modified Object::Read::prepare so it could send the epoch back out. But I don't think I have a need for it. That's the epoch for the data pool and I'm modifying the index pool. And it seems that passing in 0 works.